etl/examples/dedupe_api_schemas.py

```diff
@@ -0,0 +1,318 @@
+                    v = None
+
+            processed_record[k] = v
+
+        return processed_record
+
+    def _strip_abs_prefix(self, string):
+        return re.sub(r'^abs_', '', string, count=1)
+
+
+class MatchRequestSchema(ObjectRequestSchema):
+    threshold = fields.Float()
+
+
+class TrainRequestSchema(ObjectRequestSchema):
+    matches = fields.List(fields.Dict(), required=True)
+
+    class Decorators(object):
+        @classmethod
+        def match_processor(self, f):
+            @wraps(f)
+            def wrapper(self, object, **kwargs):
+                '''
+                Remove metadata (record ID and whether the object is a match)
+                from match objects before processing them. Add it back after.
+                '''
+                copied_object = object.copy()
+
+                cache = {
+                    'record_id': copied_object.pop('record_id', None),
+                    'match': copied_object.pop('match', None),
+                }
+
+                processed_object = f(self, copied_object, **kwargs)
+
+                for key, value in cache.items():
+                    if value is not None:
+                        processed_object[key] = value
+
+                return processed_object
+
+            return wrapper
+
+    @validates('matches')
+    def validate_matches(self, value, **kwargs):
+        match_schema = self._make_object_schema(match=True)
+
+        # If there is not a schema, skip match validation. An exception will be
+        # raised by project validation.
+        if match_schema:
+            errors = []
+
+            for match in value:
+                try:
+                    match_schema.load(match)
+                except ValidationError as e:
+                    errors.append(str(e))
+
+            if errors:
+                raise ValidationError(errors)
+
+    @post_load
+    def postprocess(self, in_data, **kwargs):
+        in_data = super().postprocess(in_data, **kwargs)
+
+        processed_matches = []
+
+        for match in in_data['matches']:
+            abstract_match = self._make_abstract_match(match)
+            processed_matches.append(self._make_processed_match(abstract_match))
+
+        in_data['processed_matches'] = processed_matches
+
+        return in_data
+
+    @Decorators.match_processor
+    def _make_abstract_match(self, object):
+        return self._make_abstract_record(object)
+
+    @Decorators.match_processor
+    def _make_processed_match(self, object):
+        return self._make_processed_record(object)
+
+
+class AddRequestSchema(ObjectRequestSchema):
+    cluster_id = UUIDString()
+
+    @validates('cluster_id')
+    def validate_cluster_id(self, value, **kwargs):
+        entity_map = 'change_log_{}'.format(self.project.id)
+
+        entity_query = '''
+            SELECT EXISTS(
+              SELECT 1 FROM "{entity_map}"
+              WHERE entity_id = :entity_id
+            )
+        '''.format(entity_map=entity_map)
```

Comment:
On second thought, I'm nervous about exposing our schema like this. I'm also skeptical that this file will be useful to devs who are not familiar with Dedupe; even with detailed knowledge of the data model and API spec, I still found it a challenging read. How about we remove this file (and squash on merging), add a link to the source code to the documentation, and add a note saying that devs who don't have access to the source can request a gist from a Dedupe.io dev? I'm expecting that the time this file will come in handy the most for me is if I need to task a dev with pulling off a tricky Marshmallow move and want to give them an example to work from, and in that case creating a gist for them based on the source code wouldn't be substantially more work than directing them to this file.